### PR TITLE
Fixes #6333: Create a periodic scheduler

### PR DESCRIPTION
**Pull request by peckpeck**
http://www.rudder-project.org/redmine/issues/6333

### CODE REVIEW

File: tree/20_cfe_basics/time_lib.cf [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+# There are 6 cases for error handling, based on:
+#  - 2 miss -> 0 job, 1 job, 2 jobs
+#  - agent overlap -> 2 jobs, 1 job
+# For which we gave 6 names:
+# - stateless:     2miss->0job & 2run->2job
+# - no_dups:       2miss->0job & 2run->1job
+# - at_least_once: 2miss->1job & 2run->2job
+# - catch_up:      2miss->1job & 2run->1job
+# - no_miss:       2miss->2job & 2run->2job
+# - exact_count:   2miss->2job & 2run->1job
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: I guess a bit more explaination of te differences would help

File: tree/20_cfe_basics/time_lib.cf [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    any::
+      "test" expression => regcmp("^\d+$", "${test_splay}");
+}
+
+# Generic Scheduler, manage periodicity and splay of execution
+#  The real start time is delayed by a constant number of minutes that is lower than the splay parameter.
+#  This splay time depends on the host properties (IP and hostname)
+#  Everything is calculated in minutes.
+#  This is the basis for other schedulers and is not made to be user friendly.
+#  The outcome are variables that can be used for final decision
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: could you list the variable we can use?

File: tree/20_cfe_basics/time_lib.cf [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    pass2.(nodups|catchup)::
+      "job_${job_id}_persist" expression => "job_${job_id}_run",
+                              persistence => "${job_periodicity}"; # shouldn't there be a -1 here ?
+
+    any::
+      "pass2" expression => "any";
+
+  methods:
+    any::
+      "any" usebundle => schedule_generic("${job_id}", "${agent_periodicity}", "${splay}", "${job_periodicity}", "${now}");
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: i fail to see why it is now that you pass as the parameter. should not it be the previous job execution ?

Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: Now is a wrong name here, as said in the comment, it is the number of minutes between now and start

File: tree/20_cfe_basics/time_lib.cf (hunk 2 above)
Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: Yes i should document it, it's job_run_id, current_run_id and max_run_id

File: tree/20_cfe_basics/time_lib.cf [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+# - stateless:     2miss->0job & 2run->2job
+# - no_dups:       2miss->0job & 2run->1job
+# - at_least_once: 2miss->1job & 2run->2job
+# - catch_up:      2miss->1job & 2run->1job
+# - no_miss:       2miss->2job & 2run->2job
+# - exact_count:   2miss->2job & 2run->1job
+
+
+# Stores now(), you can change now()'s result if you want to run tests by calling this bundle with a parameter
+bundle agent date(test_timestamp)
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: i fear this name is too generic, and can overlap with user defined code (or even one day there may be a date method in the agent).
Could you give a more "unique" name?

File: tree/20_cfe_basics/time_lib.cf [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      "timestamp" int => "${test_timestamp}";
+
+  classes:
+    any::
+      "test" expression => regcmp("^\d+$", "${test_timestamp}");
+}
+
+# Store a unique host value based on IP + hostname
+# You can change its content if you want to run tests by calling this bundle with a parameter
+bundle agent splay(test_splay)
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: same here, could you give a more unique name ?

File: tree/20_cfe_basics/time_lib.cf [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+{
+  vars:
+    any::
+      # maximum value of splay integer = 7 nibbles (2^28-1)
+      "splay_max" string => "268435455";
+      # pseudorandon deterministic hex number of 7 nibbles
+      "splay_hex" string => string_head(hash("${sys.host}${sys.ipv4}", "md5"), "7");
+      # decimal version of the splay integer
+      "splay_dec" string => execresult("/usr/bin/printf '%d' 0x${splay_hex}", "noshell");
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: splay_dec is the decimal version of the splay. Why not simply calling it "splay" ?

File: tree/20_cfe_basics/time_lib.cf [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    any::
+      "test" string => "value";
+    noerror::
+      # shift between each run = run_periodicity % agent_periodicity
+      "run_shift_real" string => eval("${job_periodicity} % ${agent_periodicity}", "math", "infix");
+      "run_shift" int => "${run_shift_real}";
+
+      # max_run_id = int(splay / agent periodicity) = min number of agent run in interval
+      "max_run_id_real" string => eval("${splay}/${agent_periodicity}", "math", "infix");
+      "max_run_id" int => "${max_run_id_real}";
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: ok, you lost me here.
the splay is 
A - the splay of the agent (like, the agent runs every 15 minutes with a splay of 3 minutes)
B - the splay of the job execution within the job interval (job every day, between 16:00 and 19:00, so a splay of 3 hours)
Either way, i can't relate it to this formula

Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: oh, it's explained. splay: on how many minutes you want to spread the runs

There are too many "splay" in this file. Could you give them more meaningful name ?

File: tree/20_cfe_basics/time_lib.cf [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      # run_id = int(splay / splay_max * max_run_id) = run id of agent at wich we must run the job
+      "job_run_id_real" string => eval("${splay.splay_dec} * ${max_run_id} / ${splay.splay_max}", "math", "infix");
+      "job_run_id" int => "${job_run_id_real}";
+
+      # elapsed in period = minutes_from_start % run periodicity = duration in minutes since the last start of the run period 
+      "elapsed" string => ifelse("null_job_periodicity", "${minutes_from_start}", eval("${minutes_from_start} % ${job_periodicity}", "math", "infix"));
+
+      # current_run_id = int(elapsed / agent_periodicity) = run id of current run
+      "current_run_id_real" string => eval("${elapsed} / ${agent_periodicity}", "math", "infix");
+      "current_run_id" int => "${current_run_id_real}";
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: I understood it ! This is great, but very hard to follow
some naming improvement would be awesome

File: tree/20_cfe_basics/time_lib.cf [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+      "splay_null" expression => strcmp("${splay}", "0");
+
+      # if splay >= run_periodicity -> error
+      # this must be checked in parent too if you have multiple periodicity
+      "job_${job_id}_error_splay_too_big" not => islessthan("${splay}", "${job_periodicity}"),
+                                          scope => "namespace",
+                                          ifvarclass => "!null_job_periodicity";
+
+      # if run_shift != 0 -> unpredictable
+      "unpredictable" not => strcmp("${run_shift}", "0");
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: Hum, isn't it the goal of it to be != 0 ?

File: tree/20_cfe_basics/time_lib.cf [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      "splay" int => "${splay_real}";
+
+      # periodicity in minutes
+      "job_periodicity_real" string => eval("${periodicity_minutes} + 60 * ${periodicity_hours} + 60 * 24 * ${periodicity_days}", "math", "infix");
+      "job_periodicity" int => "${job_periodicity_real}";
+
+      # compute time between start_* and now
+      # (min_of_day - start_minutes) + 60 * (hour of day - start hours) + 60 * 24 * (day of week - start day)%7 
+      "now_real" string => eval(strftime("localtime", "%M - ${start_on_minutes} + 60 * (%H - ${start_on_hours}) + 60 * 24 * ((%u - ${start_on_day_of_week} + 7)%%7)", "${date.timestamp}"), "math", "infix");
+      "now" int => "${now_real}";
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: now is a really poor name for that, it leads to believe that it is now, while is a difference between the start of job and now.

**Comment by ncharles** [comment]
Overall, this is great.
However, it would really benefit from renaming some variables, adding some more documentation/comments, and improve indentation on some part.

There is only one question that puzzles me, it's the "unpredictable" class, i fail to understand why it is defined as it is now

File: tree/20_cfe_basics/time_lib.cf (hunk 9 above)
Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: run_shift is a the shift between each run on the same period, which means a non 0 values will produce a delay that is always increasing

File: tree/20_cfe_basics/time_lib.cf (hunk 10 above)
Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: I'll replace it by minutes_from_start

File: tree/30_generic_methods/schedule_simple.cf [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                                                 "${periodicity_minutes}", "${periodicity_hours}", "${periodicity_days}", "${mode}");
+
+      "kept"     usebundle => _classes_success("${class_prefix}"),
+                 ifvarclass => "!job_${job_id}_run.!job_${job_id}_error";
+      "repaired" usebundle => _classes_repaired("${class_prefix}"),
+                 ifvarclass => "job_${job_id}_run";
+      "error"    usebundle => _classes_failure("${class_prefix}"),
+                 ifvarclass => "job_${job_id}_error";
+
+#      "report"                    usebundle => _logger("Scheduling ${job_id}", "${class_prefix}");
```
Comment (peckpeck) [inline tree/30_generic_methods/schedule_simple.cf]: The report was commented, uncommenting

**Comment by peckpeck** [comment]
I forgot to commit the tests, adding them

**Comment by peckpeck** [comment]
PR updated

File: tests/acceptance/30_generic_methods/schedule_simple.cf [hunk 12]
```diff
@@ -0,0 +1,129 @@
+#######################################################
+#
+# Test checking if two file permissions in a directory can be changed recursively
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/schedule_simple.cf]: the descirption is invalid

File: tests/acceptance/30_generic_methods/schedule_simple.cf [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+#      " error_null_agent_periodicity" ifvarclass => "job_${test_id}_error_null_agent_periodicity";
+#      " error_interval_too_big" ifvarclass => "job_${test_id}_error_splay_too_big";
+#      " error_interval_too_small" ifvarclass => "job_${test_id}_error_splay_too_small";
+#      " Error !" ifvarclass => "job_{test_id}_error";
+#      " Must RUN !" ifvarclass => "job_${test_id}_run";
+#      " After run ${test_id}!" ifvarclass => "job_${test_id}_after_run";
+#      " After splay ${test_id}!" ifvarclass => "job_${test_id}_after_splay";
+
+    pass3.expect_error.ok_error::
+      "TEST result OK ${test_id}";
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/schedule_simple.cf]: i really fail to see what is being tested here :/ could you explain a bit more in the code please ?

**Comment by peckpeck** [comment]
PR updated

File: tests/acceptance/30_generic_methods/schedule_simple.cf [hunk 14]
```diff
@@ -0,0 +1,138 @@
+#######################################################
+#
+# Test Tes the simple generic scheduler
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/schedule_simple.cf]: the Tes is probably a typo

File: tree/20_cfe_basics/time_lib.cf [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    pass2.!catchup::
+      "should_run" expression => strcmp("${schedule_generic.current_run_id}", "${schedule_generic.job_run_id}");
+    pass2.catchup::
+      "should_run" not => islessthan("${schedule_generic.current_run_id}", "${schedule_generic.job_run_id}");
+    pass2::
+      "job_${job_id}_run" expression => "should_run.!job_${job_id}_persist.!job_${job_id}_error",
+                          scope => "namespace";
+    pass2.(nodups|catchup)::
+      "job_${job_id}_persist" expression => "job_${job_id}_run",
+                              persistence => "${job_periodicity}"; # shouldn't there be a -1 here ?
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: only you can know if it should be -1 :)

File: tree/20_cfe_basics/time_lib.cf [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+# agent_periodicity: how often you run the agent in minutes
+# max_execution_delay: on how many minutes you want to spread the runs
+# job_periodicity: the time frame between runs in minutes 
+# minutes_from_start: the number of minutes elapsed since the first start of the job (usually: now - any_start_epoch_timestamp/60)
+#
+# If there is an error, a class is raised: job_${job_id}_error
+#    # Sample code to use in the caller 
+#      "job_${job_id}_after_delay" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.max_run_id}");
+#      "job_${job_id}_after_run" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+#      "job_${job_id}_run" expression => strcmp("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
```
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: even with this explaination, I fail to see how to use this.

File: tree/30_generic_methods/_classes_repaired.cf [hunk 17]
```diff
@@ -0,0 +1,43 @@
+#####################################################################################
+# Copyright 2013 Normation SAS
```
Comment (ncharles) [inline tree/30_generic_methods/_classes_repaired.cf]: Copyright 2015 ?

File: tree/30_generic_methods/_classes_repaired.cf [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Classes repaired definition
+# @description Defines repaired classes based on a prefix
+#
+# @parameter class_prefix      The prefix of the repaired class to defined
+#
+# @class_prefix
```
Comment (ncharles) [inline tree/30_generic_methods/_classes_repaired.cf]: shouldn't there be some text here ?

File: tree/30_generic_methods/schedule_simple_catchup.cf [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+# @parameter start_on_hours        At which hour should be the first run
+# @parameter start_on_day_of_week  At which day of week should be the first run
+# @parameter periodicity_minutes   How often should the job run
+# @parameter periodicity_hours     How often should the job run
+# @parameter periodicity_days      How often should the job run
+# @parameter mode                  "nodups": avoid duplicate runs in the same period / "catchup": avoid duplicates and one or more run have been missed, run once before next period / "stateless": no check is done on past runs
+# 
+# @class_prefix schedule_simple
+# @class_parameter job_id
+# This bundle will define a class schedule_simple_${job_id}_{kept,repaired,not_ok,ok,reached}
```
Comment (ncharles) [inline tree/30_generic_methods/schedule_simple_catchup.cf]: when is each class defined ? how should we use it ?
all information for potential user is missing :(

File: tree/30_generic_methods/schedule_simple_nodups.cf [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+# @parameter start_on_hours        At which hour should be the first run
+# @parameter start_on_day_of_week  At which day of week should be the first run
+# @parameter periodicity_minutes   How often should the job run
+# @parameter periodicity_hours     How often should the job run
+# @parameter periodicity_days      How often should the job run
+# @parameter mode                  "nodups": avoid duplicate runs in the same period / "catchup": avoid duplicates and one or more run have been missed, run once before next period / "stateless": no check is done on past runs
+# 
+# @class_prefix schedule_simple
+# @class_parameter job_id
+# This bundle will define a class schedule_simple_${job_id}_{kept,repaired,not_ok,ok,reached}
```
Comment (ncharles) [inline tree/30_generic_methods/schedule_simple_nodups.cf]: same here...

File: tree/30_generic_methods/schedule_simple_stateless.cf [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+# @parameter start_on_hours        At which hour should be the first run
+# @parameter start_on_day_of_week  At which day of week should be the first run
+# @parameter periodicity_minutes   How often should the job run
+# @parameter periodicity_hours     How often should the job run
+# @parameter periodicity_days      How often should the job run
+# @parameter mode                  "nodups": avoid duplicate runs in the same period / "catchup": avoid duplicates and one or more run have been missed, run once before next period / "stateless": no check is done on past runs
+# 
+# @class_prefix schedule_simple
+# @class_parameter job_id
+# This bundle will define a class schedule_simple_${job_id}_{kept,repaired,not_ok,ok,reached}
```
Comment (ncharles) [inline tree/30_generic_methods/schedule_simple_stateless.cf]: again, here

File: tree/20_cfe_basics/time_lib.cf (hunk 1 above)
Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: I added the explanation 6 lines above

File: tree/20_cfe_basics/time_lib.cf (hunk 8 above)
Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: I don't know how to name them better and their names are explained at the top of the method

File: tree/20_cfe_basics/time_lib.cf (hunk 15 above)
Comment (peckpeck) [inline tree/20_cfe_basics/time_lib.cf]: It was a question for you too: if we have an interval of 5mn and we persist 5 minutes, will the next run be canceled because it is withing those 5mn (with a few second error margin) ?

File: tree/30_generic_methods/_classes_repaired.cf (hunk 18 above)
Comment (peckpeck) [inline tree/30_generic_methods/_classes_repaired.cf]: There is not in similar files starting with _ like _classes_copy.cf

**Comment by peckpeck** [comment]
PR updated

File: tree/20_cfe_basics/time_lib.cf (hunk 15 above)
Comment (ncharles) [inline tree/20_cfe_basics/time_lib.cf]: Ha ! ok I understand, indeed, we should persist for job periodicity -1, otherwise the next execution will be skipped

File: tree/30_generic_methods/_classes_repaired.cf (hunk 18 above)
Comment (ncharles) [inline tree/30_generic_methods/_classes_repaired.cf]: ok

**Comment by ncharles** [comment]
@peckpeck I answered your question !

**Comment by peckpeck** [comment]
PR updated
